core/message_queue.go

```diff
@@ -0,0 +1,131 @@
+package core
+
+import (
+	"sync"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// MessageQueue stores an ordered list of messages (per actor) and enforces that their nonces form a contiguous sequence.
+// Each message is associated with a "stamp" (an opaque integer), and the queue supports expiring any list
+// of messages where the first message has a stamp below some threshold. The relative order of stamps in a queue is
+// not enforced.
+// A message queue is intended to record outbound messages that have been transmitted but not yet appeared in a block,
+// where the stamp could be block height.
+// MessageQueue is safe for concurrent access.
+type MessageQueue struct {
+	lk sync.RWMutex
+	// Message queues keyed by sending actor address, in nonce order
+	queues map[address.Address][]*queuedMessage
+}
+
+type queuedMessage struct {
+	msg   *types.SignedMessage
+	stamp uint64
+}
+
+// NewMessageQueue constructs a new, empty queue.
+func NewMessageQueue() *MessageQueue {
+	return &MessageQueue{
+		queues: make(map[address.Address][]*queuedMessage),
+	}
+}
+
+// Enqueue appends a new message for an address. If the queue already contains any messages for
+// from same address, the new message's nonce must be exactly one greater than the largest nonce
+// present.
+func (mq *MessageQueue) Enqueue(msg *types.SignedMessage, stamp uint64) error {
+	mq.lk.Lock()
+	defer mq.lk.Unlock()
+
+	q := mq.queues[msg.From]
+	if len(q) > 0 {
+		nextNonce := q[len(q)-1].msg.Nonce + 1
+		if msg.Nonce != nextNonce {
+			return errors.Errorf("Invalid nonce %d, expected %d", msg.Nonce, nextNonce)
+		}
+	}
+	mq.queues[msg.From] = append(q, &queuedMessage{msg, stamp})
+	return nil
+}
+
+// RemoveNext removes and returns a single message from the queue, if it bears the expected nonce value.
+// Returns the message removed, or nil if the queue is empty or the expected nonce was less than any
+// in the queue for that address (indicating the message had already been removed).
+// Returns an error if the expected nonce is greater than the smallest in the queue.
+// The caller may wish to check that the returned message is equal to that expected (not just in nonce value).
+func (mq *MessageQueue) RemoveNext(sender address.Address, expectedNonce uint64) (*types.SignedMessage, error) {
```

Comment:
It's not clear how the caller is going to know the expected nonce.
In general, there's a lot of logic around nonce and block height tracking that I was expecting to see inside this queue. Will this data structure be wrapped in another, smarter message queue looking thing that handles that logic? It seems a bit much for a plumbing struct.

Reply:
The caller is going to see a message in a block and read its nonce from that. Nonces should be contiguous, so a gap will error here. But the sequence may also restart from an earlier point after a reorg.